enhancements/microshift/microshift-pluggable-core-components.md

```diff
@@ -0,0 +1,213 @@
+  * `/usr/lib/microshift/manifests.d/user-app-1/`
+  * `/usr/lib/microshift/manifests.d/user-app-2/`
+  * `/etc/microshift/manifests.d/user-app-3/`
+  * `...`
+
+* Split the RPM packages to resemble those optional components.
+
+* Preserve previous behavior in terms of RPM package installation.
+
+### Non-Goals
+
+* Making DNS component optional.
+* Making the certificate management component optional.
+
+## Proposal
+
+To allow modularity of the described core components, the following changes
+would be introduced:
+
+Update the service manager to allow core components as kustomization yamls,
+in the same way we already support for user application manifests. i.e.:
+
+`/usr/lib/microshift/components.d/networking-ovn/kustomize.yaml`
+`/usr/lib/microshift/components.d/storage-topolvm/kustomize.yaml`
+`/usr/lib/microshift/components.d/openshift-router/kustomize.yaml`
+
+  MicroShift component manager would scan the `/usr/lib/microshift/components.d/`
+  directory and extract metadata from the yamls to analyze dependencies between
+  core components at boot. Then start the right componments in the right order.
+
+* The kustomize component loader would be refactored so it can be used for this
+  purpose. We should be able to pre-process kustomize yamls to insert template
+  variables as we do today for core assets.
+
+The following changes would be introduced to packaging:
+
+* Convert `microshift` into a virtual package that depends on `microshift-core`,
+  `microshift-networking-ovn`, `microshift-storage-topolvm`, `microshift-openshift-router`
+  to preserve the current behavior.
+
+* Convert the `microshift-networking` package would be made a virtual package which
+  depends on `microshift-networking-ovn` to preserver current behavior.
+
+* The `microshift-networking-ovn` package would include the necessary
+  `/usr/lib/microshift/components.d/networking-ovn/` yamls, including crds.
+
+* The `microshift-storage-topolvm` package would include the necessary
+  `/usr/lib/microshift/components.d/storage-topolvm` yamls, including crds.
+
+* The `microshift-openshift-router` package would include the necessary
+  `/usr/lib/microshift/components.d/openshift-router` yamls, including crds.
+
+The rebase script should be updated to extract those core components
+as kustomize yamls instead of embedded assets.
+
+### Workflow Description
+
+#### Deploying
+
+The workflow for deployment is already described in the user histories.
+
+#### Upgrading
+
+Upgrading optional components would behave in the same way as upgrading
+components where the assets are embedded into the MicroShift binary.
+
+#### Configuring
+
+Configuration of the optional components is done by package selection
+when the images are created.
+
+#### Deploying Applications
+
+Applications are deployed as usual, we would add support for multiple
+directories in the user component manifests as described in the 2nd goal.
+
+### API Extensions
+
+None.
+
+### Risks and Mitigations
+
```

Comment:
We need to think how we'll handle lifecycle of pluggable components and how they'll work with upgrades.

- Let say I deployed virtual `microshift` RPM, everything's good and stuff.
- Now I want to upgrade, so I shut down microshift, containers keep running, I install newer RPMs and start microshift again.
- We just go in and apply everything that's in there?
- What if we have a direct conflict and something cannot be applied or the opposite - second instance is deployed (I'm just speculating here :D)
- What if some CRD matures, changes version - do we assume it's on user (or us for our components) to resolve that manually - migrate the CRs to newer version (or use mutating webhook which might be a waste of resources to run longer than migration in case of MicroShift).
- What if user is happy with removing everything and deploying newer version - unless it's saved somewhere, old component version would have to be removed first (using files, because with newer RPMs they might not match 1 to 1), microshift stopped, RPM installed, microshift start to deploy newer version.
- And if the yamls contain Go template syntax, I think kubectl will fail to parse them so the won't be usable even to run `kubectl kustomize . | kubectl delete -f -`

Reply:
Not sure if here or in design doc but it would be nice to have some diagrams regarding pluggable components lifecycle

Reply:
> We need to think how we'll handle lifecycle of pluggable components and how they'll work with upgrades.
> 
>     * Let say I deployed virtual `microshift` RPM, everything's good and stuff.
> 
>     * Now I want to upgrade, so I shut down microshift, containers keep running, I install newer RPMs and start microshift again.

I think that should not be supported. As it's not supported to change the node name for similar reasons.

> 
>     * We just go in and apply everything that's in there?
> 
>     * What if we have a direct conflict and something cannot be applied or the opposite - second instance is deployed (I'm just speculating here :D)

Yes it's a valid concern, something that we would need to disallow or document as not supported. It would be a problem that we would experience with the main or alternative proposals as well.

> 
>     * What if some CRD matures, changes version - do we assume it's on user (or us for our components) to resolve that manually - migrate the CRs to newer version (or use mutating webhook which might be a waste of resources to run longer than migration in case of MicroShift).

That's a good point. Static code enabled/disabled by config or plugins would allow to do in-place migrations before boot for example. 

> 
>     * What if user is happy with removing everything and deploying newer version - unless it's saved somewhere, old component version would have to be removed first (using files, because with newer RPMs they might not match 1 to 1), microshift stopped, RPM installed, microshift start to deploy newer version.
 
I believe that use case should not be supported under any architecture. As for the name name, once adopted it should not be supported to switch. There are many details to tests an many things which could go wrong, many system details which could have been wired by the previous plugin (network settings, network plumbing, mounts, volumes...)

IMO a migration path should only be offered if we deprecate a component and customers are forced to switch, and may be not even that. Customers could still have "data save procedures" (for volumes and content) before a switch which would re-initialized the device completely to the new configuration and then reload the data.




> 
>     * And if the yamls contain Go template syntax, I think kubectl will fail to parse them so the won't be usable even to run `kubectl kustomize . | kubectl delete -f -`

I was planning to handle that by pre-processing the yamls, for example: applying go templating by rendering new files in /tmp/${TMP_DIR}, executing them, removing them.

Reply:
> I think that should not be supported. As it's not supported to change the node name for similar reasons.
What if someone does it, we say it's not supported and then we'll get question: how do I fix it?

> I was planning to handle that by pre-processing the yamls, for example: applying go templating by rendering new files in /tmp/${TMP_DIR}, executing them, removing them.
But that works from the code, what about the user?

Another scenario: (upstream) user/dev wants to migrate from one CNI to the other. I know this is extreme example, edging on madness, but it's still possible. Even if we don't support it, someone will ask how to do it and if not, they'll do it on their own.

I know I ask edge case questions, but I we think about them right now, they might spare us some headaches later on (and I'm assuming here full pluggability, not the limited one in 4.13)

Reply:
There are those corner cases, for sure, and we'll need to list and triage them.
As guideline for such triage, I suggest trying to keep things simple and drawing some lines in the sand early on. For example, I don't think CNI upgrades are a realistic scenario for MicroShift and not even for OpenShift in the general case (i.e. any CNI), but just OSDN-->OVNk.

Reply:
I think we shall keep basic options open for upstream when drawing the line here.
A basic scenario is that:
1) user installs microshift-core and starts the service, the cluster is fully up and running with default CNI plugin
2) user wants to install another CNI instead of using the default

Reply:
In that case they can use the cleanup-all-microshift-data and reboot to the new plugin.